### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by arthurWD**

### CODE REVIEW

File: lib/linked_rails/iri_mapper.rb [hunk 1]
```diff
@@ -3,57 +3,119 @@
...
+        single_resource_from_iri(iri, user_context) || raise(ActiveRecord::RecordNotFound)
+      end
+
+      def single_resource_from_opts(opts, user_context)
+        opts[:class]&.requested_single_resource(opts[:params], user_context) unless opts[:params] == 'index'
+      end
+
+      private
+
+      def verify_iri!(iri)
```
Comment (rescribet) [inline lib/linked_rails/iri_mapper.rb]: Meaning can be better expressed: `ensure_absolute_iri!`

File: lib/linked_rails/iri_mapper.rb [hunk 2]
```diff
@@ -3,57 +3,119 @@
...
 
-        ApplicationRecord.descendants.detect { |m| m.to_s == type.classify } ||
-          ApplicationRecord.descendants.detect { |m| m.to_s == type.split('/').last.classify }
+      def resource_from_opts(opts, user_context)
+        opts[:class]&.requested_resource(opts, user_context)
       end
 
-      def resource_action?(action)
-        %w[show update destroy].include?(action)
+      def route_params_to_opts(params) # rubocop:disable Metrics/MethodLength
```
Comment (rescribet) [inline lib/linked_rails/iri_mapper.rb]: - Hopefully this is short enough for rubocop
- Doesn't modify incoming argument
- Better readability
- I've assumed `controller` and `action` must be removed from the returned object under the `params` key.

```ruby
def route_params_to_opts(params)
    opts = params.dup
    controller = opts.delete(:controller)
    controller_class = class_for_controller(controller)

    return EMPTY_IRI_OPTS.dup if controller_class.blank?

    {
        action: opts.delete(:action),
        class: controller_class,
        params: opts
    }.with_indifferent_access
end

private

def class_for_controller(controller)
    controller&.classify&.safe_constantize ||
        "::#{controller.camelize}Controller"
            .safe_constantize
            .try(:controller_class)
end
```

File: lib/linked_rails/model/iri_mapping.rb [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            resource&.singular_resource = true
+            resource
+          else
+            requested_single_resource(opts[:params], user_context)
+          end
+        end
+
+        def requested_single_resource(params, _user_context)
+          if self < ActiveRecord::Base
+            find_by(id: params[:id]) if params.key?(:id)
```
Comment (rescribet) [inline lib/linked_rails/model/iri_mapping.rb]: Seems like this should use [`primary_key`](https://api.rubyonrails.org/classes/ActiveRecord/AttributeMethods/PrimaryKey/ClassMethods.html#method-i-primary_key)

File: lib/linked_rails/routes.rb [hunk 4]
```diff
@@ -15,7 +15,68 @@ def use_linked_rails(opts = {}) # rubocop:disable Metrics/AbcSize
...
+      get '(*parent_iri)/actions', to: 'actions/items#index'
+      get '(*parent_iri)/actions/:id', to: 'actions/items#show'
+      get '(*parent_iri)/new', to: 'actions/items#show', id: :create
+      get '(*parent_iri)/edit', to: 'actions/items#show', id: :update
+      get '(*parent_iri)/delete', to: 'actions/items#show', id: :destroy
+      get '(*parent_iri)/trash', to: 'actions/items#show', id: :trash
+      get '(*parent_iri)/untrash', to: 'actions/items#show', id: :untrash
+    end
+
+    def linked_resource(klass, controller: nil, nested: false, path: nil)
```
Comment (rescribet) [inline lib/linked_rails/routes.rb]: We only use `klass` and `path` in apex, and the usages of `path` seem like they could be moved to the model as well, so perhaps most of these args can be omitted.

File: lib/linked_rails/routes.rb [hunk 5]
```diff
@@ -15,7 +15,68 @@ def use_linked_rails(opts = {}) # rubocop:disable Metrics/AbcSize
...
+      route_key = path || klass.route_key
+
+      linked_routes(klass, route_key, controller: controller, nested: nested) do
+        yield if block_given?
+      end
+    end
+
+    def singular_linked_resource(klass, controller: nil, nested: true, path: nil)
+      route_key = path || klass.singular_route_key
```
Comment (rescribet) [inline lib/linked_rails/routes.rb]: Always check for `klass.singular_route_key` in the execution path of `linked_resource` and omit this method?

File: lib/linked_rails/routes.rb [hunk 6]
```diff
@@ -15,7 +15,68 @@ def use_linked_rails(opts = {}) # rubocop:disable Metrics/AbcSize
...
+      linked_routes(klass, route_key, controller: controller, nested: nested, singular_route: true) do
+        yield if block_given?
+      end
+    end
+
+    private
+
+    # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/MethodLength, Metrics/PerceivedComplexity
+    def linked_routes(klass, path, singular_route: nil, controller: nil, nested: nil)
+      route_name ||= klass.name.demodulize.tableize
```
Comment (rescribet) [inline lib/linked_rails/routes.rb]: `||=` on a local variable, is this a hidden variable on module level?

File: lib/linked_rails/routes.rb [hunk 7]
```diff
@@ -15,7 +15,68 @@ def use_linked_rails(opts = {}) # rubocop:disable Metrics/AbcSize
...
+
+      linked_routes(klass, route_key, controller: controller, nested: nested, singular_route: true) do
+        yield if block_given?
+      end
+    end
+
+    private
+
+    # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/MethodLength, Metrics/PerceivedComplexity
+    def linked_routes(klass, path, singular_route: nil, controller: nil, nested: nil)
```
Comment (rescribet) [inline lib/linked_rails/routes.rb]: Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/MethodLength, Metrics/PerceivedComplexity

```ruby
    def active_controller_opts(route_options)
      {
        controller: route_options[:controller],
        only: %i[show],
        path: route_options[:routes_path],
      }
    end

    def creatable(klass)
      klass.enhanced_with?(LinkedRails::Enhancements::Creatable, :Routing)
    end

    def install_route(klass, path, controller: nil, nested: nil)
      opts = route_options(klass, path, controller, nested)

      send('resources', opts[:route_name], active_controller_opts(opts), route_block(klass))

      post(parentable_path, to: "#{controller}#create") if creatable(klass)
      get(parentable_path, to: "#{controller}#index")
    end

    def install_singular_route(klass, path, controller: nil, nested: nil)
      opts = route_options(klass, path, controller, nested)
      ac_opts = active_controller_opts(opts).merge(singular_route: true)

      send('resource', opts[:route_name], ac_opts, route_block(klass))

      post(routes_path, to: "#{controller}#create", singular_route: true) if creatable(klass)
    end

    def route_block(klass)
      -> do
        include_route_concerns(klass: klass)

        yield if block_given?
      end
    end

    def route_options(klass, path, controller: nil, nested: nil)
      {
        route_name: klass.name.demodulize.tableize,
        controller: klass.name.tableize,
        parentable_path: "(*parent_iri)/#{path}",
        routes_path: nested ? parentable_path : path,
      }.with_indifferent_access
    end
```

**Review by rescribet — CHANGES_REQUESTED** [review]

File: lib/linked_rails/routes.rb (hunk 4 above)
Comment (arthurWD) [reply lib/linked_rails/routes.rb]: These keys are used by linked_rails-auth

**Pull request merged**